File: bukkit/src/main/java/me/devnatan/inventoryframework/pipeline/ItemClickInterceptor.java
package me.devnatan.inventoryframework.pipeline;

import me.devnatan.inventoryframework.VirtualView;
import me.devnatan.inventoryframework.component.Component;
import me.devnatan.inventoryframework.component.ItemComponent;
import me.devnatan.inventoryframework.context.SlotClickContext;
import org.bukkit.event.inventory.InventoryClickEvent;
import org.bukkit.event.inventory.InventoryType;
import org.jetbrains.annotations.NotNull;

/**
 * Intercepted when a player clicks on an item the view container.
 */
public final class ItemClickInterceptor implements PipelineInterceptor<VirtualView> {

    @Override
    public void intercept(@NotNull PipelineContext<VirtualView> pipeline, @NotNull VirtualView subject) {
        if (!(subject instanceof SlotClickContext)) return;

        final SlotClickContext context = (SlotClickContext) subject;
        final InventoryClickEvent event = context.getClickOrigin();
        if (event.getSlotType() == InventoryType.SlotType.OUTSIDE) return;

        final Component component = context.getComponent();
        if (component == null) return;

        if (component instanceof ItemComponent) {
            final ItemComponent item = (ItemComponent) component;

            // inherit cancellation so we can un-cancel it
            context.setCancelled(item.isCancelOnClick());
        }

        if (component.getInteractionHandler() != null)
            component.getInteractionHandler().clicked(component, context);

        event.setCancelled(context.isCancelled());
    }
}

File: inventory-framework-platform-minestom/src/main/kotlin/me/devnatan/inventoryframework/pipeline/ItemClickInterceptor.kt
package me.devnatan.inventoryframework.pipeline

import me.devnatan.inventoryframework.VirtualView
import me.devnatan.inventoryframework.component.ItemComponent
import me.devnatan.inventoryframework.context.SlotClickContext
import net.minestom.server.event.inventory.InventoryPreClickEvent

/** Intercepted when a player clicks on an item the view container. */
class ItemClickInterceptor : PipelineInterceptor<VirtualView> {
    override fun intercept(
        pipeline: PipelineContext<VirtualView>,
        subject: VirtualView,
    ) {
        if (subject !is SlotClickContext) return

        val event: InventoryPreClickEvent = subject.clickOrigin
        event.inventory ?: return

        val component = subject.component ?: return

        if (component is ItemComponent) {
            val item: ItemComponent = component

            // inherit cancellation so we can un-cancel it
            subject.isCancelled = item.isCancelOnClick
        }
    }
}
